src/discrete_rv.jl

```diff
@@ -78,3 +78,69 @@ function Base.rand!(out::AbstractArray{T}, d::DiscreteRV) where T<:Integer
 end
 
 @deprecate draw Base.rand
+
+
+struct MVDiscreteRV{TV1<:AbstractArray,TV2<:AbstractVector,K,TI<:Integer}
+    q::TV1
+    Q::TV2
+    dims::NTuple{K,TI}
+
+    function MVDiscreteRV{TV1,TV2,K,TI}(q::TV1, Q::TV2, dims::NTuple{K,TI}) where {TV1,TV2,K,TI}
+        abs(sum(q) - 1.0) > 1e-10 && error("q should sum to 1")
+        abs(Q[end] - 1.0) > 1e-10 && error("Q[end] should be 1")
+        length(Q) != prod(dims) && error("Number of elements is inconsistent")
+
+        new{TV1,TV2,K,TI}(q, Q, dims)
+    end
+end
+
+
+function MVDiscreteRV(q::TV1) where TV1<:AbstractArray
+    Q = cumsum(vec(q))
+    dims = size(q)
+
+    return MVDiscreteRV{typeof(q),typeof(Q),length(dims),eltype(dims)}(q, Q, dims)
+end
+
+
+"""
+Make a single draw from the multivariate discrete distribution.
+
+##### Arguments
+
+- `d::MVDiscreteRV`: The `MVDiscreteRV` type represetning the distribution
+
+##### Returns
+
+- `out::NTuple{Int}`: One draw from the discrete distribution
+"""
+function Base.rand(d::MVDiscreteRV)
+    x = rand()
+    i = searchsortedfirst(d.Q, x)
+
+    return ind2sub(d.dims, i)
+end
+
+"""
+Make multiple draws from the discrete distribution represented by a
+`MVDiscreteRV` instance
+
+##### Arguments
+
+- `d::MVDiscreteRV`: The `DiscreteRV` type representing the distribution
+- `k::Int`
+
+##### Returns
+
+- `out::Vector{NTuple{Int}}`: `k` draws from `d`
+"""
+Base.rand(d::MVDiscreteRV{T1,T2,K,TI}, k::V) where {T1,T2,K,TI,V} =
```

Comment:
What is `k` supposed to be? I think it should have `k::Integer` instead of a free type param